text/0000-proc-macros.md

```diff
@@ -0,0 +1,417 @@
+the annotated item until after macro expansion. That would allow for more
+flexible macros, since the input would not be constrained to Rust syntax. However,
+this would require identifying items from tokens, rather than from the AST, which
+would require additional rules on token trees and may not be possible.
+
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+### macros with an extra identifier
+
+We currently allow procedural macros to take an extra ident after the macro name
+and before the arguments, e.g., `foo! bar(...)` where `foo` is the macro name
+and `bar` is the extra identifier. This is used for `macro_rules` and is useful
+for macros which define classes of items, rather than instances of items. E.g.,
+a `struct!` macro might be used similarly to the `struct` keyword.
+
+My feeling is that this macro form is not used enough to justify its existence.
+From a design perspective, it encourages uses of macros for language extension,
+rather than syntactic abstraction. I feel that such macros are at higher risk of
+making programs incomprehensible and of fragmenting the ecosystem).
+
+Therefore, I would like to remove them from the language. Alternatively, they
+could be incorporated into the new design by having another kind of macro
+function:
+
+```
+#[macro_with_ident]
+pub fn foo(&Token, TokenStream, &mut MacroContext) -> TokenStream;
+```
+
+where the first argument is the extra identifier.
+
+
+### Linking model
+
+Currently, procedural macros are dynamically linked with the compiler. This
+prevents the compiler being statically linked, which is sometimes desirable. An
+alternative architecture would have procedural macros compiled as independent
+programs and have them communicate with the compiler via IPC.
+
+This would have the advantage of allowing static linking for the compiler and
+would prevent procedural macros from crashing the main compiler process.
+However, designing a good IPC interface is complicated because there is a lot of
+data that might be exchanged between the compiler and the macro.
+
+I think we could first design the syntax, interfaces, etc. and later evolve into
+a process-separated model (if desired). However, if this is considered an
+essential feature of macro reform, then we might want to consider the interfaces
+more thoroughly with this in mind.
+
+
+### Interactions with constant evaluation
+
+Both procedural macros and constant evaluation are mechanisms for running Rust
+code at compile time. Currently, and under the proposed design, they are
+considered completely separate features. There might be some benefit in letting
+them interact.
+
+
+### Inline procedural macros
+
+It would nice to allow procedural macros to be defined in the crate in which
+they are used, as well as in separate crates (mentioned above). This complicates
+things since it breaks the invariant that a crate is designed to be used at
+either compile-time or runtime. I leave it for the future.
```

Comment:
That wouldn't really be allowed, I wouldn't think. It's hard enough to make it work before expansion, *during* expansion it would be a nightmare.
*Unless* we had "macro modules" which were split off as an entire crate, disregarding anything else in the current crate (i.e. they would be the root of their own crate). But that might get a bit confusing.